go/vt/mysqlctl/xtrabackupengine.go

```diff
@@ -0,0 +1,401 @@
+/*
+Copyright 2019 The Vitess Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mysqlctl
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+
+	"github.com/klauspost/pgzip"
+	"vitess.io/vitess/go/mysql"
+	"vitess.io/vitess/go/vt/dbconfigs"
+	"vitess.io/vitess/go/vt/logutil"
+	"vitess.io/vitess/go/vt/mysqlctl/backupstorage"
+	"vitess.io/vitess/go/vt/vterrors"
+)
+
+// XtrabackupEngine encapsulates the logic of the xtrabackup engine
+// it implements the BackupEngine interface and contains all the logic
+// required to implement a backup/restore by invoking xtrabackup with
+// the appropriate parameters
+type XtrabackupEngine struct {
+}
+
+var (
+	// path where backup engine program is located
+	// TODO assume lives on path and remove this option?
+	xtrabackupEnginePath = flag.String("xtrabackup_root_path", "", "directory location of the xtrabackup executable, e.g., /usr/bin")
+	// TBD whether to support flags to pass through to backup engine
+	xtrabackupBackupFlags = flag.String("xtrabackup_backup_flags", "", "flags to pass to backup command. these will be added to the end of the command")
+	// TBD whether to support flags to pass through to restore phase
+	xtrabackupRestoreFlags = flag.String("xtrabackup_restore_flags", "", "flags to pass to restore command. these will be added to the end of the command")
+	// streaming mode
+	xtrabackupStreamMode = flag.String("xtrabackup_stream_mode", "tar", "which mode to use if streaming, valid values are tar and xbstream")
+)
+
+const (
+	streamModeTar      = "tar"
+	xtrabackup         = "xtrabackup"
+	xbstream           = "xbstream"
+	binlogInfoFileName = "xtrabackup_binlog_info"
+	successMsg         = "completed OK!"
+)
+
+func (be *XtrabackupEngine) backupFileName() string {
+	fileName := "backup"
+	if *xtrabackupStreamMode != "" {
+		fileName += "."
+		fileName += *xtrabackupStreamMode
+	}
+	if *backupStorageCompress {
+		fileName += ".gz"
+	}
+	return fileName
+}
+
+// ExecuteBackup returns a boolean that indicates if the backup is usable,
+// and an overall error.
+func (be *XtrabackupEngine) ExecuteBackup(ctx context.Context, cnf *Mycnf, mysqld MysqlDaemon, dbconfigs *dbconfigs.DBConfigs, logger logutil.Logger, bh backupstorage.BackupHandle, backupConcurrency int, hookExtraEnv map[string]string) (bool, error) {
+
+	/* TODO uncomment if we keep this option
+	if *xtrabackupEnginePath == "" {
```

Comment:
I wouldn't enforce this. Only fail if below the executable can't be found. Basically this TODO should be enough: 
```
// TODO check that the executable file exists, exit if it doesn't or isn't executable
```